File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/EXÁMENES/examen_22-23_extraordinaria/deepspace_rb/PowerEfficientSpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStation'
require_relative 'PowerEfficientSpaceStationToUI'
require_relative 'Transformation'

module Deepspace

    #Encapsula la mejora de una Space Station a PowerEfficientSpaceStation
    class PowerEfficientSpaceStation < SpaceStation

        #Potencia extra de disparo y proteccion.
        @@EFFICIENCYFACTOR=1.1

        #param: station --> SpaceStation
        def initialize (station)
            newCopy(station)  #Si ponemos super(station) llamaría al initialize de SpaceStation, que recibe otros parametros
        end

        #Interfaz

        def getUIversion
            return (PowerEfficientSpaceStationToUI.new(self))
        end

        def fire()
            @@EFFICIENCYFACTOR*super
        end

        def protection()
            @@EFFICIENCYFACTOR*super
        end

        #Se permite la transformacion a BetaPowerEfficientSpaceStation,
        #pero no a SpaceCity.
    
        def setLoot(loot)
        
            transformation=super(loot)
            
            if(transformation==Transformation::SPACECITY) 
                transformation= Transformation::NOTRANSFORM
            end
            return transformation
        end
        

    end
end

File: INFORMÁTICA/PROGRAMACIÓN Y DISEÑO ORIENTADO A OBJETOS/PRÁCTICAS/PROYECTOS/deepspace/deepspace/src/deepspace/PowerEfficientSpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author leandrofdez
 */

//Encapsula la mejora de una Space Station a PowerEfficientSpaceStation.
class PowerEfficientSpaceStation extends SpaceStation{
    
    //Potencia extra de disparo y proteccion.
    private static final float EFFICIENCYFACTOR=1.1f;
    
    
    
    public PowerEfficientSpaceStation(SpaceStation station){
         super(station);
        
    }

    
    
    /*
    Añade un factor extra a la potencia disparo.
    */
    @Override
    public float fire(){
        
        return EFFICIENCYFACTOR*super.fire();
    }
    
    /*
    Añade un factor extra a la potencia de escudo.
    */
    @Override
    public float protection(){
        
        return EFFICIENCYFACTOR*super.protection();
    }
    
    
    /*
    Se permite la transformacion a BetaPowerEfficientSpaceStation,
    pero no a SpaceCity.
    */
    @Override
    public Transformation setLoot(Loot loot){
        
        Transformation transformation= super.setLoot(loot);
        
        if(transformation==Transformation.SPACECITY) 
            transformation= Transformation.NOTRANSFORM;
        
        return transformation;
    }
    
    //Interfaz
    @Override
    public PowerEfficientSpaceStationToUI getUIversion(){
        return (new PowerEfficientSpaceStationToUI(this));
    }
    
    
}
